mirrord-layer/src/pod_api.rs

```diff
@@ -412,8 +392,345 @@ async fn create_job_pod_agent(
+    async fn runtime_data(&self, client: &Client, namespace: &str) -> Result<RuntimeData>;
+}
+
+impl Target {
+    pub async fn runtime_data(&self, client: &Client, namespace: &str) -> Result<RuntimeData> {
+        match self {
+            Target::Pod(pod) => pod.runtime_data(client, namespace).await,
+            Target::Deployment(deployment) => deployment.runtime_data(client, namespace).await,
+        }
+    }
+}
+
+impl FromStr for DeploymentData {
+    type Err = LayerError;
+
+    fn from_str(input: &str) -> Result<Self> {
+        let target_data = input.split('/').collect::<Vec<&str>>();
+        match target_data.first() {
+            Some(&"deployment") if target_data.len() == 2 => Ok(DeploymentData {
+                deployment: target_data[1].to_string(),
+            }),
+            _ => Err(LayerError::InvalidTarget(format!(
+                "Provided target: {:?} is not a deployment.",
+                input
+            ))),
+        }
+    }
+}
+
+#[derive(Debug, Clone, PartialEq)]
+pub(crate) struct PodData {
+    pub pod_name: String,
+    pub container_name: Option<String>,
+}
+
+impl FromStr for PodData {
+    type Err = LayerError;
+
+    fn from_str(input: &str) -> Result<Self> {
+        let target_data = input.split('/').collect::<Vec<&str>>();
+        match target_data.first() {
+            Some(&"pod") if target_data.len() == 2 => Ok(PodData {
+                pod_name: target_data[1].to_string(),
+                container_name: None,
+            }),
+            Some(&"pod") if target_data.len() == 4 && target_data[2] == "container" => {
+                Ok(PodData {
+                    pod_name: target_data[1].to_string(),
+                    container_name: Some(target_data[3].to_string()),
+                })
+            }
+            _ => Err(LayerError::InvalidTarget(format!(
+                "Provided target: {:?} is neither a pod or a deployment.",
+                input
+            ))),
+        }
+    }
+}
+
+#[async_trait]
+impl RuntimeDataProvider for PodData {
+    async fn runtime_data(&self, client: &Client, namespace: &str) -> Result<RuntimeData> {
+        let pod_api: Api<Pod> = Api::namespaced(client.clone(), namespace);
+        let pod = pod_api.get(&self.pod_name).await?;
+        let (
+            ContainerRuntime {
+                container_runtime,
+                socket_path,
+                container_id,
+            },
+            container_name,
+        ) = || -> Option<(String, String)> {
+            let container_statuses = &pod.clone().status?.container_statuses?;
+            let container_name = if let Some(container_name) = self.container_name.clone() {
+                container_name
+            } else {
+                pod.clone()
+                    .spec?
+                    .containers
+                    .first()
+                    .map(|container| container.name.clone())?
+            };
+            let container_info = container_statuses
```

Comment:
`container_info` is called `runtime_and_id` a few rows down, why the inconsistency?
(Also `container_info` isn't a good name - how's it different from `container_data`, which is used for completely different things elsewhere in the code?)